File: app/src/main/java/com/meta4projects/takenote/database/typeconverters/SubsectionConverter.java
package com.meta4projects.takenote.database.typeconverters;

import androidx.room.TypeConverter;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.meta4projects.takenote.models.Subsection;

import java.lang.reflect.Type;
import java.util.ArrayList;

public class SubsectionConverter {

    @TypeConverter
    public String fromSubsection(ArrayList<Subsection> subsections) {
        if (subsections == null) {
            return null;
        }

        Gson gson = new Gson();
        Type type = new TypeToken<ArrayList<Subsection>>() {
        }.getType();
        return gson.toJson(subsections, type);
    }

    @TypeConverter
    public ArrayList<Subsection> toSubsection(String value) {
        if (value == null) {
            return null;
        }

        Gson gson = new Gson();
        Type type = new TypeToken<ArrayList<Subsection>>() {
        }.getType();
        return gson.fromJson(value, type);
    }
}

File: app/src/main/java/com/meta4projects/takenote/database/typeconverters/SubsectionConverter.kt
package com.meta4projects.takenote.database.typeconverters

import androidx.room.TypeConverter
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.meta4projects.takenote.database.Lists

class SubsectionConverter {
    @TypeConverter
    fun fromSubsection(subsections: Lists.Subsections?): String? {
        if (subsections == null) return null
        val gson = Gson()
        val type = object : TypeToken<Lists.Subsections?>() {}.type
        return gson.toJson(subsections, type)
    }

    @TypeConverter
    fun toSubsection(value: String?): Lists.Subsections? {
        if (value == null) return null
        val gson = Gson()
        val type = object : TypeToken<Lists.Subsections?>() {}.type
        return gson.fromJson(value, type)
    }
}